### PR TITLE
Add GetBalance to CoinView

### PR DESCRIPTION
**Pull request by quantumagi**
Merge PR #110 first.

Implements a very fast `IEnumerable<(uint height, long satoshis)> GetBalance(TxDestination txDestination);`

The first enumeration value comes from the `Balance` table and subsequent values are adjusted via the `BalanceAdjustment` table.

Only a fraction of a second (~40 ms) response time was observed to return ~1800 balance adjustments all the way to genesis.

Implemented ONCE in `Coindb<T>` which replaces `LevelDbCoindb` and `RocksDbCoindb`.

The code deals with the following issues:
- Keeping track of the height up to which balances are indexed so that address indexing can be toggled on/off without resulting in gaps.
- Replacing `AddressIndexer` with `AddressIndexerCV` and respecting the `-addressindex` setting.
- Adding missing XML comments.
- Adding the `IDb` compliant wrapper classes `RocksDb` and `LevelDb`.
```
/// <summary>
/// Returns a combination of (height, satoshis) values with the cumulative balance up to the corresponding height.
/// </summary>
/// <param name="txDestination">The destination value derived from the address being queried.</param>
/// <returns>A combination of (height, satoshis) values with the cumulative balance up to the corresponding height.</returns>
/// <remarks>Balance updates (even when nett 0) are delivered for every height at which transactions for the address had been recorded 
/// and as such the returned heights can be used in conjunction with the block store to discover all related transactions.</remarks>
IEnumerable<(uint height, long satoshis)> GetBalance(TxDestination txDestination);
```

Instead of adding migration code the coin database is automatically truncated and re-built (+- 5 minutes).

### CODE REVIEW

**Comment by quantumagi** [comment]
Still testing.

File: src/Stratis.Bitcoin.Features.Consensus/CoinViews/Coindb/LeveldbCoindb.cs [hunk 1]
```diff
@@ -304,48 +299,80 @@ public HashHeightPair Rewind(HashHeightPair target)
...
                 for (int height = startHeight; height > (target?.Height ?? (startHeight - 1)) && height > (startHeight - MaxRewindBatchSize); height--)
                 {
-                    byte[] row = this.leveldb.Get(new byte[] { rewindTable }.Concat(BitConverter.GetBytes(height).Reverse()).ToArray());
+                    byte[] rowKey = BitConverter.GetBytes(height).Reverse().ToArray();
+                    byte[] row = this.leveldb.Get(rewindTable, rowKey);
 
                     if (row == null)
                         throw new InvalidOperationException($"No rewind data found for block at height {height}.");
 
-                    batch.Delete(BitConverter.GetBytes(height));
```
Comment (quantumagi) [inline src/Stratis.Bitcoin.Features.Consensus/CoinViews/Coindb/LeveldbCoindb.cs:320]: The table prefix is missing here.

File: src/Stratis.Bitcoin.Features.Consensus/CoinViews/Coindb/LeveldbCoindb.cs [hunk 2]
```diff
@@ -7,22 +7,19 @@
...
+using Stratis.Bitcoin.Interfaces;
 using Stratis.Bitcoin.Utilities;
 
 namespace Stratis.Bitcoin.Features.Consensus.CoinViews
 {
-    /// <summary>
-    /// Persistent implementation of coinview using the dBreeze database engine.
-    /// </summary>
-    public class LevelDbCoindb : ICoindb, IStakedb, IDisposable
+/// <summary>
```
Comment (zeptin) [inline src/Stratis.Bitcoin.Features.Consensus/CoinViews/Coindb/LeveldbCoindb.cs:15]: Can be neatened up a bit

Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Consensus/CoinViews/Coindb/LeveldbCoindb.cs]: Fixed the indentation issue.

**Comment by fassadlr** [comment]
Hi Gustav, release 1.3 will go out tomorrow as we are already in RC with the masternodes. Can you please push this to

https://github.com/stratisproject/StratisFullNode/tree/release/1.3.1.0

File: src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinviewHelper.cs [hunk 3]
```diff
@@ -20,8 +20,10 @@ public OutPoint[] GetIdsToFetch(Block block, bool enforceBIP30)
             {
                 if (enforceBIP30)
                 {
+                    // Calculate the hash outside the loop.
```
Comment (zeptin) [inline src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinviewHelper.cs:23]: I vaguely recall we cache the hash once it is computed for the first time for a transaction. But there is no harm structuring the loop in this more sensible way.

File: src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexerCV.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        }
+
+        public AddressBalancesResult GetAddressBalances(string[] addresses, int minConfirmations = 0)
+        {
+            return new AddressBalancesResult()
+            {
+                Balances = addresses.Select(address => new AddressBalanceResult()
+                {
+                    Address = address,
+                    Balance = new Money(this.coinView.GetBalance(AddressToDestination(address)).First().satoshis)
```
Comment (zeptin) [inline src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexerCV.cs:56]: I guess it doesn't make much difference for current use cases, but what if `AddressToDestination` returns null for scriptPubKey formats that don't fall into one of the recognised categories?

File: src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinviewHelper.cs (hunk 3 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinviewHelper.cs:23]: I did actually have a look but could not find any caching. The reason could be that there is no cut-off really to establish when a transaction had been completely built.

File: src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexerCV.cs (hunk 4 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexerCV.cs]: Fixed.

**Comment by quantumagi** [comment]
See #1031

**Pull request closed without merging**